File: android/src/main/java/expo/modules/contenttransition/glyphs/GlyphBlur.kt
package expo.modules.contenttransition.glyphs

import android.os.Build
import androidx.compose.ui.graphics.BlurEffect
import androidx.compose.ui.graphics.RenderEffect
import androidx.compose.ui.graphics.TileMode
import kotlin.math.roundToInt

internal object GlyphBlur {
  // Whole-pixel buckets make a decaying blur step down in visible notches, so the glyph reads as
  // vibrating while it settles. Eighth-pixel buckets sit below that threshold and keep the cache small.
  private const val STEPS_PER_PX = 8f
  private const val MIN_RADIUS = 0.375f
  private const val MAX_CACHED = 512

  private val cache = HashMap<Int, RenderEffect>()

  fun effect(radius: Float): RenderEffect? {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S || radius < MIN_RADIUS) {
      return null
    }

    val bucket = (radius * STEPS_PER_PX).roundToInt()
    cache[bucket]?.let { return it }

    if (cache.size >= MAX_CACHED) {
      cache.clear()
    }

    val quantized = bucket / STEPS_PER_PX
    return BlurEffect(quantized, quantized, TileMode.Decal).also { cache[bucket] = it }
  }
}

File: web/glyphs/GlyphBlur.ts
const STEPS_PER_PX = 8;
const MIN_RADIUS = 0.375;
const MAX_CACHED = 512;

const SIGMA_SCALE = 0.57735;
const SIGMA_BIAS = 0.5;

const NO_FILTER = 'none';

const cache = new Map<number, string>();

const sigma = (radius: number): number => (radius > 0 ? radius * SIGMA_SCALE + SIGMA_BIAS : 0);

const filter = (radius: number): string => {
  if (!(radius >= MIN_RADIUS)) return NO_FILTER;

  const bucket = Math.round(radius * STEPS_PER_PX);
  const cached = cache.get(bucket);
  if (cached) return cached;

  if (cache.size >= MAX_CACHED) cache.clear();

  const value = `blur(${sigma(bucket / STEPS_PER_PX).toFixed(3)}px)`;
  cache.set(bucket, value);
  return value;
};

const GlyphBlur = { NO_FILTER, sigma, filter } as const;

export { GlyphBlur };
